sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -138,15 +138,17 @@ namespace Azure { namespace Core {
     }
 
     /**
-     * @brief Get a value associated with a \p key parameter within this context or the branch of
-     * contexts this context belongs to.
+     * @brief Try to get a value associated with a \p key parameter within this context or the
+     * branch of contexts this context belongs to.
      *
      * @param key A key associated with a context to find.
+     * @param outputValue A reference to the value corresponding to the key to be set, if found
+     * within the context tree.
      *
-     * @return A value associated with the context found; an empty value if a specific value can't
-     * be found.
+     * @return If found, returns true, with outputValue set to the value associated with the context
+     * found; otherwise returns false.
      */
-    template <class T> const T& Get(const std::string& key) const
+    template <class T> bool TryGetValue(const std::string& key, T& outputValue) const
```

Comment:
@CaseyCarter, in this case, would you recommend accepting an `outputValue` parameter as a reference (`T&`) or an explicit pointer (`T*`)?

I don't want someone to pass in a null pointer.

Does `T&&` makes sense here at all?

Reply:
> Does T&& makes sense here at all?

It doesn't make sense to me.

Reply:
> I don't want someone to pass in a null pointer.

It makes sense to me to allow passing null pointer. In which case, the user just wants to check if the key exists and the type matches.

Reply:
> In which case, the user just wants to check if the key exists and the type matches.

That's an interesting idea. However, then the API behavior changes if the user passes in `null` vs non-null, which is confusing. 
- `null` ptr - always returns true or false, **false if type mismatch**
- non-null ptr - sometimes returns true or false, **abort if type mismatch** - what should TryGetValue do/return in this case?

I would prefer if the behavior was always consistent.

Reply:
I'm thinking

* null ptr - returns true if key exists, else false. Type mismatch will throw or abort.

Reply:
I prefer `T&`

Reply:
That behavior is equivalent to the `HasKey` method, which we are removing because:
- there is no real scenario for it, where the caller wouldn't then try to get the actual value after that returns true anyway
- Using HasKey then GetValue encourages a poor coding pattern where we do two look-ups within the context tree

Can you think of a real-world use case or scenario where the user would want to pass in a null ptr to `TryGetValue` (or previously call `HasKey`) that isn't followed by making another call to get the value?

Reply:
For example, the size of value is big and copy can be expensive. The key may not exist. In the past user can do

```C++
if (context.HasKey("key")) {
    const ValueType& val = context.GetValue("key");
}
```

Now there's no way to do that.

An option is to throw exception from `GetValue` if key doesn't exist.

Reply:
well @Jinming-Hu we would use a raw-pointer or smart-ptr in order to avoid the copy and just access to it
like
```cpp
BigStructure value;
context c = otherContext.WithValue(key, &value);

// Then to read it from context
auto ref = c.TryGetValue<BigStructure*>(key);
```

Reply:
@vhvb1989 Yeah, that's an option.